### PR TITLE
Basic Playerclass Struct 

### PR DESCRIPTION
**Pull request by ethanwtai**
I created the files `class.h`, `class_structs.h`, and `class.c` which contain the new, init, and free functions for the playerclass module. Tests for these are also added in the `test_class.c` file. I stubbed out classes for stats, skilltrees, combat actions, and noncombat actions. Progress was documented in [#552](https://github.com/uchicago-cs/chiventure/issues/552).

### CODE REVIEW

**Comment by ethanwtai** [comment]
I've fixed the formatting and documentation issues you brought up. I've also changed the testing file to include a "class-creating function" to reduce redundant variables.

File: include/playerclass/class.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+/* 
+ * Allocates memory for a new player class. Only creates a deep copies of the 
+ * strings name, short description, and long description.
+ * 
+ * Parameters:
+ *  - name, shortdesc, longdesc: Name and descriptions of the class
+ *  - attr: the attributes of the class
+ *  - stat: the stats of the class
+ *  - battle: the battle/combat actions of the class
+ *  - action: the noncombat actions of the class
```
Comment (dwahme) [inline include/playerclass/class.h:20]: Both of these should be `skill_t`'s. Additionally, I think naming the attributes `combat` and `noncombat` would be better names

File: include/playerclass/class.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * Initializes values for a player class. Only creates a deep copies of the 
+ * strings name, short description, and long description.
+ * 
+ * Parameters:
+ *  - class: a pointer to the class to be initialized
+ *  - name, shortdesc, longdesc: Name and descriptions of the class
+ *  - attr: the attributes of the class
+ *  - stat: the stats of the class
+ *  - battle: the battle/combat actions of the class
+ *  - action: the noncombat actions of the class
```
Comment (dwahme) [inline include/playerclass/class.h:40]: Both of these should be `skill_t`'s. Additionally, I think naming the attributes `combat` and `noncombat` would be better names

File: include/playerclass/class_structs.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+// combat actions stub
+typedef struct battle {
+    obj_t* battle;
+} battle_t;
+
+//noncombat actions stub
+typedef struct action {
+    obj_t* action;
+} action_t;
```
Comment (dwahme) [inline include/playerclass/class_structs.h:38]: Combine these two stubs and name them `skill_t` instead

File: include/playerclass/class_structs.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#define MAX_LONG_DESC_LEN (300)
+
+// playerstats stub
+typedef struct stats {
+    obj_t* stat;
+} stats_t;
+
+// skilltrees stub
+typedef struct skilltree {
+    obj_t* skill;
```
Comment (dwahme) [inline include/playerclass/class_structs.h:27]: This attribute should be named `skilltree`

File: include/playerclass/class_structs.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    stats_t* stats;
+
+    // Class skilltree
+    skilltree_t* skill;
+
+    // Class combat actions
+    battle_t* battle;
+
+    // Class noncombat actions
+    action_t* action;
```
Comment (dwahme) [inline include/playerclass/class_structs.h:65]: Change this to `skill_t *noncombat`

File: include/playerclass/class_structs.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    obj_t* attributes;
+
+    // All the stats of the class
+    stats_t* stats;
+
+    // Class skilltree
+    skilltree_t* skill;
+
+    // Class combat actions
+    battle_t* battle;
```
Comment (dwahme) [inline include/playerclass/class_structs.h:62]: Change this to `skill_t *combat`

File: include/playerclass/class_structs.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    char* longdesc;
+
+    // An object containing all the attributes of the class
+    obj_t* attributes;
+
+    // All the stats of the class
+    stats_t* stats;
+
+    // Class skilltree
+    skilltree_t* skill;
```
Comment (dwahme) [inline include/playerclass/class_structs.h:59]: Rename to `skilltree`

File: src/playerclass/src/class.c [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+/* See class.h */
+class_t* class_new(char* name, char* shortdesc, char* longdesc,
+                   obj_t* attr, stats_t* stat, skilltree_t* skill,
+                   battle_t* battle, action_t* action)
+{
+    int rc;
+    class_t* c = (class_t*) malloc(
+        sizeof(class_t)
+    );
```
Comment (dwahme) [inline src/playerclass/src/class.c:23]: This should all be on one line. Also, you should never use `malloc()`- instead, always use `calloc()`. It's much safer

File: src/playerclass/src/class.c [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                   obj_t* attr, stats_t* stat, skilltree_t* skill,
+                   battle_t* battle, action_t* action)
+{
+    if (class == NULL)
+    {
+        fprintf(stderr, "Class to initialize in class_init is NULL.\n");
+        return EXIT_FAILURE;
+    }
+
+    class->name = (char*) calloc(MAX_NAME_LEN, sizeof(char));
```
Comment (dwahme) [inline src/playerclass/src/class.c:51]: This should be `MAX_NAME_LEN + 1` to account for the null terminator

File: src/playerclass/src/class.c [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    if (class->name == NULL)
+    {
+        fprintf(stderr, "Could not allocate memory for name in class_init.\n");
+        return EXIT_FAILURE;
+    }
+    strncpy(class->name, name, MAX_NAME_LEN);
+    class->shortdesc = (char*) calloc(MAX_SHORT_DESC_LEN, sizeof(char));
```
Comment (dwahme) [inline src/playerclass/src/class.c:58]: Add a newline before this line. Also, this should be `MAX_SHORT_DESC_LEN + 1`

File: src/playerclass/src/class.c [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    if (class->name == NULL)
+    {
+        fprintf(stderr, "Could not allocate memory for short description in "
+                        "class_init.\n");
+        return EXIT_FAILURE;
+    }
+    strncpy(class->shortdesc, shortdesc, MAX_SHORT_DESC_LEN);
+    class->longdesc = (char*) calloc(MAX_LONG_DESC_LEN, sizeof(char));
```
Comment (dwahme) [inline src/playerclass/src/class.c:66]: Newline and `MAX_LONG_DESC_LEN + 1`

File: src/playerclass/src/class.c [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    if (class->name == NULL)
+    {
+        fprintf(stderr, "Could not allocate memory for long description in "
+                        "class_init.\n");
+        return EXIT_FAILURE;
+    }
+    strncpy(class->longdesc, longdesc, MAX_LONG_DESC_LEN);
+    class->attributes = attr;
```
Comment (dwahme) [inline src/playerclass/src/class.c:74]: Put a newline here

**Comment by ethanwtai** [comment]
I've fixed some formatting and collapsed the battle_t and action_t structs into one skill_t struct as per @dwahme 's previous comments.

**Comment by dwahme** [comment]
Issue Score: ✔️+

Comments:
Great job! Here are a few things you may want to improve in future issues:

The title of an issue has to convey what the issue will be producing. In the future, try using simple titles with a "Verb-Object" structure that clearly states what you will be doing in the issue (e.g., “Design and implement a ‘circle’ module”, “Research libraries for drawing simple geometric shapes”, etc.)

**Pull request merged**